### ISSUE TITLE
provide NDK integration for CMake as part of the NDK package

### ISSUE THREAD

**Issue by DanAlbert** [android/ndk#463]
@bradking @rcdailey @kravindran

Tracking bug for getting the NDK's CMake support integrated with the stuff built in to CMake.

Based on a chat we had with @bradking, the right approach here is to teach CMake to load most of the knowledge that is currently in our cmake toolchain from the CMake side of things. If CMake is building for the NDK and it finds those files in the NDK (it won't for NDK releases prior to this being fixed), it will import those NDK modules to allow us to override CMake's rules (so we can add features that don't require a new CMake release).

We'll still need a CMake toolchain file in the NDK that uses these plugins for the foreseeable future since we still need to support versions of CMake older than whatever version this new implementation makes it into.

https://android-review.googlesource.com/c/395513/ was the first stab at implementing this. It's definitely bitrotted by now, and it still needs the CMake side of the work to be done (not sure if @kravindran had some WIP commits for this, but I didn't find them).

### EDIT 22 Jan 2019

The most up to date plan for this is to ship the Platform/*.cmake files that CMake uses to support Android as part of the NDK itself. The toolchain file will continue to exist as a way to interface to this new implementation. When using an older version of CMake, the toolchain file will need to disable policy CMP0017 (hopefully that's the only workaround that we'll need).

This gives the NDK the control it needs to work around toolchain bugs and update default behaviors while also reducing the friction caused by the toolchain file that does much more than toolchain files were ever intended to do.

**Comment by Zingam** [comment]
Now that AS 3 supports external CMake and hopefully will work in Beta 8 will you prioritize completing the CMake integration?

**Comment by DanAlbert** [comment]
@kravindran, any update on this?

**Comment by alexcohn** [comment]
What about 'teaching' cmake to use **Android.mk**? I mean, let it invoke **ndk-build** for sub-modules.

**Comment by Zingam** [comment]
This issue should be a priority really. I hope that it won't get postponed too much again for some reason.

**Comment by alexcohn** [comment]
Too begin with, the should be easily discoverable instructions on how to build something for Android with Cmake from command line (i.e. not through Android Studio).

**Comment by alexcohn** [comment]
Note that https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android-with-the-ndk is quite outdated (fits NDK r14 or earlier).

**Comment by rcdailey** [comment]
How do you know it's the settings being out of date as opposed to just lack of support for newer NDKs? Again this goes back to the fact that CMake has to be changed each time the NDK changes. If no one is doing that actively (which it doesn't seem they are) then this is expected. That's not an issue of documentation, it's an issue of continuity of support.

**Comment by Zingam** [comment]
@DanAlbert Our project supports 1, 2, 3, 4, 5... etc. current and even last generation platforms and we are less developers than people who comment here. So every little hiccup with the tools piles up and becomes a burden.
There is this platform vendor S, whose tools are implemented in such way that all you need to do is: 1. Install the SDK, 2. Provision the DevKit, 3. Start a new project - and everything just works. Which is an example that things could be done properly. Then there are other vendors like M, which everytime whenever they release a new SDK everything falls apart.

I really would prefer to work on developing an end-user product than being a build systems and tools virtuoso.

**Comment by enh** [comment]
> I really would prefer to work on developing an end-user product than being a build systems and tools virtuoso.

indeed. that's probably true for everyone :-)

we are trying to get to the point where we're closer to "company S" than "company M" on the spectrum, but unfortunately we're having to dig our way out of a lot of historical problems. part of the big idea behind NDK r19 is that at that point all (or at least most) of the configuration disappears into the toolchain.

with r18 we're pretty close: there's one compiler, one libc++, one set of headers... r19 moves a lot of the build system configuration into the toolchain (which, sadly, will cause disruption for anyone with a custom build system, but should mean that that's the _last_ time we disrupt them in that way). more critically for this specific bug, r19 is when we plan to work with upstream cmake so that they "just work" with r19 and future NDKs.

longer-term there might be some more disruption (we currently have different linkers for different architectures, and we'd like to get everyone on to lld), but again, that's disruption in the name of long-term calm.

sorry that you have to live through the turbulent times, but hopefully it's a consolation that the very purpose of the disruption is to move to the "company S" kind of world :-)

**Comment by Zingam** [comment]
In that context if external CMake is used in Android Studio why is it necessary to set the version of CMake:
```cmake {
            ...
            version "3.7.1"
        }
```
If CMake can do that already: `cmake_minimum_required(VERSION 3.6.0)`.

**Comment by Zingam** [comment]
@rcdailey I know that to use an external CMake I need to specify "version" but I don't see why instead of using a Boolean value it is necessary to type in a specific unportable number.

**Comment by rcdailey** [comment]
Ah I see what you mean. Yes, I had the same issue, because the "version"
attribute does not follow the same semantics as `cmake_minimum_required()`
(the gradle attribute enforces an exact match, while CMake only requires a
minimum, so it's less restrictive). Gradle's functionality should not
interfere or supersede any functionality in CMake from my standpoint.

I filed a bug about this issue on the Google bug tracker here:
https://issuetracker.google.com/issues/110693527

I encourage you to contribute a comment to that issue. I had also been
emailing Jomo at Google directly to discuss this, since he's responsible
for the CMake integration functionality of the gradle plugin. You may
consider opening another issue for better visibility since he's since
closed the one I linked above.

If I had to speculate, I think the version might be used for other things
besides just which CMake to use. For example, he uses CMake's server mode
in newer versions of the plugin so he may need to know the version for
backward compatibility functionality of the plugin itself. But then I'd
question why he isn't querying the version from CMake directly for that.
Again, this is speculative, and he may not even be doing something that
sophisticated with the version number.

We definitely need more people to complain about this to Google. So far I
think it's just been me!

---------
Robert Dailey

**Comment by rcdailey** [comment]
> looking at that bug, it looks like there's new syntax so you can say "3.7+" to mean `>= 3.7`... isn't that sufficient?

I thought so too, until Jomo replied after that with:

> When we add a more recent version to the SDK then that version will be used (if installed)

So basically if the internal CMake within the SDK is upgraded, "3.7+" won't mean "use external CMake" anymore. And this would be a functional issue, if say, in the `build.gradle` it was "3.7+" (to force usage of external CMake) and in `CMakeLists.txt` it specified `cmake_minimum_required(VERSION 3.12)` and they upgraded the bundled CMake to something like version 3.9. CMake would all of a sudden fail because the minimum required CMake version is no longer met, even though a viable candidate version is on `PATH` in the host system.

**Comment by alexcohn** [comment]
If you ask me, I don't see why explicit choice of external CMake may be important. It's understandable that somebody may choose a the newest CMake that is not (yet) bundled with Android SDK. Or decide to save disk space and choose not to download CMake with AS SDK Manager. But if you specify "3.7+", why should it matter which copy of compliant CMake is picked up?

**Comment by rcdailey** [comment]
> But if you specify "3.7+", why should it matter which copy of compliant CMake is picked up?

@alexcohn As I explained in my previous reply, the version it picks up might meet the minimum version requirements specified by `build.gradle`, but not the minimum required specified by the `CMakeLists.txt`. Please review the detailed example I gave earlier, I feel it explains the issue pretty well.

On to @DanAlbert's response: I apologize as technically mentioning the Gradle issues here is off topic. I got off on a tangent starting with @Zingam's question related to Gradle. I think it's worth discussing, but the NDK developers probably won't be able to help with that as it's an issue specific to Android Studio & Gradle, as far as I see it.

> The problem that I've run in to is that I don't see a way to maintain those 4 assumptions without making problem 1 worse. If we want to support new and old versions of CMake, as well as toolchain and non-toolchain interfaces to each, then the problem goes from toolchain vs not toolchain vs not X cmake new enough vs not.

I like your idea of defining a cut-off point. We already know that even though NDK support exists in CMake, it wasn't really ready. You and I have discussed a couple of issues in the past where the toolchain just operates superiorly to CMake. There are actual issues that appear when using CMake's Android support that go away when switching to the toolchain file. I mean, given this, we don't want to encourage people to use it, especially since we can't go back in time and make it work better. So: 

1. A specific version of CMake (let's say 3.13, for the sake of an example) is "blessed" by the NDK team to represent the first release of CMake with acceptable/stable NDK support
1. Versions of CMake 3.13 and greater are expected to support versions of the NDK back to a NDK-team-defined minimum NDK version (you used r11 as an example I think).
1. Using the toolchain file with CMake 3.13 and greater should issue an error, and a diagnostic that points the user to the Android Platform page of the CMake docs to assist in the transition (you could make this a warning if you prefer to deprecate it for a time).
1. Versions of CMake prior to 3.13 should use the toolchain file, not the built in support. You unfortunately won't be able to enforce this, the user will just have to notice something isn't working right.

Also worth noting that I think "updating cmake" might not be as invasive as it sounds (in the case where you're worried people can't update it). From what I saw in the CMake code, all the Android platform stuff is stored in several CMake scripts provided with the installation. So if it's easier for them to update that in isolation, or perhaps provide their own (which CMake also defines a structure for) then that may make it a non-issue.

This presents an interesting challenge to the CMake developers, where it might make more sense to allow distribution of cross compilation platform support scripts independently of CMake itself. In other words, allowing groups of CMake scripts under `Modules/Platform` to be bundled and distributed outside of CMake itself. This might give you the best of both worlds:

1. CMake still controls the platform toolchain behavior
1. The owners/maintainers of the platform toolchain logic can have more ownership & control over those scripts, and distribute them to older versions of CMake (assuming the toolchain scripts themselves are backward compatible)

**Comment by DanAlbert** [comment]
> Using the toolchain file with CMake 3.13 and greater should issue an error, and a diagnostic that points the user to the Android Platform page of the CMake docs to assist in the transition (you could make this a warning if you prefer to deprecate it for a time).

This is something I'm fairly certain we can't do. You don't even have to look outside this thread to find people who think the best thing we could do to the NDK is not change the interface :)

> This presents an interesting challenge to the CMake developers, where it might make more sense to allow distribution of cross compilation platform support scripts independently of CMake itself. In other words, allowing groups of CMake scripts under Modules/Platform to be bundled and distributed outside of CMake itself.

The implementation I have actually does more or less this. The files in Modules/Platform just delegate to the same files that live in the NDK. It seems to work pretty well.

> Actually I have an interest in not using the toolchain either, because of a few reasons:
>
> 1. I don't have to worry about the path to the NDK in my build scripts (right now I need it to point CMake to the toolchain file).
> 2. I don't have to make copies of the toolchain script and use it locally (instead of the NDK provided one) in cases where bugs come up

Isn't that still the case using the built-in stuff? CMake still needs to know where the NDK is, and you'll still need to locally modify *something* if you want to get early bug fixes.

> 3. Overall I just find CMake's built in support a little easier to understand and use, although most options aren't that much different (ABI, Min API, things like that).

Anything we can improve in the toolchain file in the mean time?

**Comment by Zingam** [comment]
@rcdailey summed it up brilliantly.


> > Using the toolchain file with CMake 3.13 and greater should issue an error, and a diagnostic that points the user to the Android Platform page of the CMake docs to assist in the transition (you could make this a warning if you prefer to deprecate it for a time).
> 
> This is something I'm fairly certain we can't do. You don't even have to look outside this thread to find people who think the best thing we could do to the NDK is not change the interface :)

You just removed GCC!

**Comment by stephenhines** [comment]
> @rcdailey summed it up brilliantly.
> 
> > > Using the toolchain file with CMake 3.13 and greater should issue an error, and a diagnostic that points the user to the Android Platform page of the CMake docs to assist in the transition (you could make this a warning if you prefer to deprecate it for a time).
> > 
> > 
> > This is something I'm fairly certain we can't do. You don't even have to look outside this thread to find people who think the best thing we could do to the NDK is not change the interface :)
> 
> You just removed GCC!

Yes, it was removed after a long period of being deprecated. There was more than sufficient notice for the removal of GCC for all NDK users. It isn't realistic to support multiple distinct toolchains.

**Comment by Zingam** [comment]
I am sorry. My intention was never to spark a discussion and waste your time.

**Comment by DanAlbert** [comment]
I do need answers to my questions above though. This bug is blocked until I can understand what people are hoping to get out of it.

**Comment by rcdailey** [comment]
> > Using the toolchain file with CMake 3.13 and greater should issue an error, and a diagnostic that points the user to the Android Platform page of the CMake docs to assist in the transition (you could make this a warning if you prefer to deprecate it for a time).
> 
> This is something I'm fairly certain we can't do. You don't even have to look outside this thread to find people who think the best thing we could do to the NDK is not change the interface :)

Yeah, and I get that, which is why I parenthetically mentioned you could make it a warning. Similar to how you handled the GCC removal, you could share that strategy here: Give people ample time to move their build scripts over to the CMake-defined set up.

I think this transition is very important because of documentation. Ideally you will want people to turn to the [official CMake toolchain page](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android-with-the-ndk) as the singular, one-stop shop for information on using CMake with the NDK. I mean, I feel strongly about this, since I'm one of the folks affected by the confusion of having multiple ways of doing the same thing.

> > This presents an interesting challenge to the CMake developers, where it might make more sense to allow distribution of cross compilation platform support scripts independently of CMake itself. In other words, allowing groups of CMake scripts under Modules/Platform to be bundled and distributed outside of CMake itself.
> 
> The implementation I have actually does more or less this. The files in Modules/Platform just delegate to the same files that live in the NDK. It seems to work pretty well.

Maybe not appropriate to go into the details here, but I'm curious: Isn't it as simple as adding your own path to the module search directory and using the same structure, it will find those files first instead of the ones bundled with CMake?

> > Actually I have an interest in not using the toolchain either, because of a few reasons:
> > 
> > 1. I don't have to worry about the path to the NDK in my build scripts (right now I need it to point CMake to the toolchain file).
> > 2. I don't have to make copies of the toolchain script and use it locally (instead of the NDK provided one) in cases where bugs come up
> 
> Isn't that still the case using the built-in stuff? CMake still needs to know where the NDK is, and you'll still need to locally modify _something_ if you want to get early bug fixes.

So with builtin CMake support for the NDK, it automatically searches your environment (plus a couple of other rules) for the path to the NDK. That makes the NDK path management unnecessary in my actual build scripts (I can just invoke CMake without providing it an explicit path, assuming I follow the rules CMake set forth).

Right now, I have to specify the path to the NDK's toolchain script as an argument to CMake itself. This is not implicitly done, as it would be with CMake's built in support.

It's a minor detail, but improves the usage a little bit.

And to my second point: That "something" that I modify shouldn't be my build scripts. Basically, if you fix a bug in the Android platform modules, and I don't want to wait on a new NDK release to get those, I should be able to copy just those scripts and drop them into a directory somewhere to take precedence over the ones supplied by the current (older) NDK. Note that this is an edge case and I wouldn't expect you to design for this case (obviously we don't want bugs in the first place). I thought it was worth mentioning, since I had personally been affected by this scenario just a couple of times.

> > 1. Overall I just find CMake's built in support a little easier to understand and use, although most options aren't that much different (ABI, Min API, things like that).
> 
> Anything we can improve in the toolchain file in the mean time?

I'm not sure why you are interested in improving it. Could you give me an example of what you expect to improve? I know my question is probably going to create an infinite loop, but really: If the strategy is to get away from the toolchain in favor of a CMake-centric solution, I think the energy should be focused on that and the toolchain goes into maintenance mode (i.e. you only make changes out of necessity to fix bugs, instead of quality-of-life improvements). I apologize if I'm misunderstanding.

**Comment by rcdailey** [comment]
I wanted to leave my last paragraph as a new comment since it's not related to @DanAlbert's last response. I really just want to clarify (in a simpler/shorter way) the importance of this effort:

Simplicity and unification of NDK support with CMake.

That's it. Documentation issues are just a symptom of the real problem. Right now we have 2 ways of integrating NDK and CMake together:

1. NDK's provided toolchain
2. CMake's Android platform support

We don't want two. I pick #2 because it's more coupled to CMake itself, which means we can make the integration work better (since you have more control in that case than you do with a simple, singular toolchain file).

With this problem solved, you get everything else naturally: Documentation will be a non-issue, as people will have 1 place to go for answers (the CMake docs page). I think once you get past the deprecation strategy for this, this is the long term place I want to see the NDK and CMake be from an end user standpoint.

**Comment by DanAlbert** [comment]
> Maybe not appropriate to go into the details here, but I'm curious: Isn't it as simple as adding your own path to the module search directory and using the same structure, it will find those files first instead of the ones bundled with CMake?

Didn't know that was an option tbh. We're not CMake users ourselves, so don't know most of its features,  so thanks for the pointer!

It looks like this won't work because of a policy in `include` though:

> There is one exception to this: if the file which calls include() is located itself in the CMake builtin module directory, then first the CMake builtin module directory is searched and CMAKE_MODULE_PATH afterwards.

I think that's not an issue as long as we're careful to include a (probably empty) file for all of the possible CMake files that might be included from CMake's C++ source? Might be more fragile than the alternative, but I'll play with it.

> it automatically searches your environment

Ahhhh okay. This is the part I was missing. Thanks for the explanation.

> And to my second point: That "something" that I modify shouldn't be my build scripts. Basically, if you fix a bug in the Android platform modules, and I don't want to wait on a new NDK release to get those, I should be able to copy just those scripts and drop them into a directory somewhere to take precedence over the ones supplied by the current (older) NDK. Note that this is an edge case and I wouldn't expect you to design for this case (obviously we don't want bugs in the first place). I thought it was worth mentioning, since I had personally been affected by this scenario just a couple of times.

The change to your build files being that you need to update the path to the toolchain file? In a world where you're using the built-in support you'd be able to avoid any changes because you already need to modify `CMAKE_MODULE_PATH` and would just drop modifications in to that directory? I'm not a CMake user so I'm performing a fair amount of guesswork to try to figure out how things like this would actually get used :)

That makes sense to me and gives me a better idea of what folks want the UX to look here. I think I can work with this.

> I'm not sure why you are interested in improving it. Could you give me an example of what you expect to improve?

That's what I'm asking you :) If there are rough edges that we can trivially smooth out until the built-in support is ready we might as well. Wasn't thinking I'd set this aside in favor of major toolchain file changes, but if there's a default that ought to be inferred or an easier way to specify something that's basically a one liner.

> Right now we have 2 ways of integrating NDK and CMake together:
> 
> 1. NDK's provided toolchain
> 2. CMake's Android platform support
> 
> We don't want two.

Trust me, it's a giant PITA for us too. What I'm leaning toward here is making CMake's built in (although delegated back to the NDK) the only actual *implementation*, but I think we will be keeping around the toolchain file because it is a different interface and there are some things that I think are more in the realm of NDK policy defaults (think things like `-Werror=format-security`; that's part of the legacy of the toolchain file that ours is based on) than things that should be in the CMake implementation. I think if you're using the built-in CMake support it ought to be as unopinionated as possible (like it is for every other platform CMake supports), but if you do want the NDK to continue imposing its opinions on your build (or maybe just don't want to deal with migration) you can keep using the toolchain file. It would internally use the built-in support, but it would pass a bunch of extra `CMAKE_CXX_FLAGS` by default, essentially. I'm not picturing keeping anything more complicated than that.

That said, I too would really like to have only one option. The prevailing opinion in this thread seems to be that we ought to just kill the toolchain, but I don't think a sample of a handful of people really tells me enough to know if I can get away with that :)

**Comment by rcdailey** [comment]
Again, and sorry for repeating myself, I think the long term goal *is* to kill the toolchain. Otherwise we can't rely on the CMake documentation. This needs to be 100% CMake. The only things your team should be responsible for are:

1. NDK structure and things specific to the NDK itself, regardless of build system
2. Contributions back to the `Platform/Modules` scripts in the CMake project relevant to the Android platform.

That's it. The road to get there will be a long one, with lots of strategic deprecation of things (including the toolchain!). I don't see any reasonable middle ground. The challenge is going to be how to approach this transition, and how to easily educate users on the transition itself, while giving them time to do so.

The GCC remark was a surprisingly relevant mention: It's a good analogy for what you need to do here. While the 2 are completely unrelated, conceptually the difficulty and challenges are similar: How to force your users to change something.

Making the toolchain file a simple wrapper for CMake itself is definitely a way to deprecate the feature, and help users transition. But long term, as the absolute final stop, I would expect `rm android.toolchain.cmake` to happen.

**Comment by DanAlbert** [comment]
I suppose I should stop arguing with people that want me to do less work :)

I think what I'll do is like you've suggested: sprinkle deprecation warnings in the toolchain to encourage people to switch. I'm probably being overly cautious. It's annoying to have to fiddle with your build scripts to adapt, but it's not actually difficult. Worst case scenario we find out I was right and have to keep it, but we might well assume we can do more until we know otherwise.

Thanks again for the detailed feedback; it really helps. I'll see what I can come up with and post back here when I have something that's ready for a test drive in case you're feeling adventurous.

**Comment by DanAlbert** [comment]
> > Maybe not appropriate to go into the details here, but I'm curious: Isn't it as simple as adding your own path to the module search directory and using the same structure, it will find those files first instead of the ones bundled with CMake?
> 
> Didn't know that was an option tbh. We're not CMake users ourselves, so don't know most of its features, so thanks for the pointer!
> 
> It looks like this won't work because of a policy in `include` though:
> 
> > There is one exception to this: if the file which calls include() is located itself in the CMake builtin module directory, then first the CMake builtin module directory is searched and CMAKE_MODULE_PATH afterwards.
> 
> I think that's not an issue as long as we're careful to include a (probably empty) file for all of the possible CMake files that might be included from CMake's C++ source? Might be more fragile than the alternative, but I'll play with it.

Yes, https://cmake.org/cmake/help/git-master/policy/CMP0017.html prevents this. When I asked Brad if he had any suggestions on how we could make an implementation like this work, he rejected the implementation idea I've proposed above. CMake won't approve any change that delegates *work* outside of CMake, so the only thing we're able to put in the NDK is plain old data. We also can't touch `CMAKE_MODULE_PATH`.

> And to my second point: That "something" that I modify shouldn't be my build scripts. Basically, if you fix a bug in the Android platform modules, and I don't want to wait on a new NDK release to get those, I should be able to copy just those scripts and drop them into a directory somewhere to take precedence over the ones supplied by the current (older) NDK. Note that this is an edge case and I wouldn't expect you to design for this case (obviously we don't want bugs in the first place). I thought it was worth mentioning, since I had personally been affected by this scenario just a couple of times.

I'm not sure how to accommodate this given the above. It sounds to me like this is something CMake specifically wants to prevent (nothing should be able to accidentally override CMake behavior, and automatically loading things from the user's tree can cause that).

**Comment by rcdailey** [comment]
Does Brad then expect to own the knowledge & maintenance burden of supporting NDK 100%? And if so, what if that causes issues with NDK release schedule and priority needs? I'm not sure what the alternative is.

Just to make sure I understand: You are saying the CMake developers essentially do not want platform CMake scripts (or any bundled scripts, for that matter) to be maintained externally to and independent of CMake itself?

**Comment by alexcohn** [comment]
I will hopefully expand this later, but in the nutshell, I believe that if the 'external' CMake could export the variables expected by 'integrated' CMake, the same scripts that run today in Android Studio, will work with any CMake, won't need the toolchain file and nobody will see the difference (for some definition of 'nobody').

I am rather sceptical about the advantages of dropping the toolchain file for easier application of bug fixes.

**Comment by alexcohn** [comment]
@Zingam, I agree that looking at `cmake_minimum_required` should be enough. I don't want an extra flag in gradle, be it number or number with +, or boolean.

Also, I wonder: how much do you need to modify your CMake scripts to support Android along with "S" and "M"? My experience with Android/Linux/Windows cross-platform is that the two have very little in common after all, and it's easier to keep separate build scripts per platform. And we use XCode native build system for relevant targets to make the life a bit trickier.

**Comment by rcdailey** [comment]
@alexcohn: Probably a discussion for another thread. It was my fault the first time we got derailed on that discussion, so again I apologize for making the middle of the conversation misleading.

**Comment by DanAlbert** [comment]
> Just to make sure I understand: You are saying the CMake developers essentially do not want platform CMake scripts (or any bundled scripts, for that matter) to be maintained externally to and independent of CMake itself?

I don't want to put words in his mouth, but he's on this thread so he can speak for himself (@bradking, because I don't think he's actively following here).

My understanding though (I was much more concise here than his email was) is that everything about the layout and variables used in `$CMAKE/Modules/Platform` is an implementation detail and subject to change at any time. It's not so much that they don't *want* externally maintained platforms, but rather that they do not currently have a stable interface to enable that.

> I am rather sceptical about the advantages of dropping the toolchain file for easier application of bug fixes.

I'm starting to wonder as well. I took a quick run through our toolchain file to get a handle on what information we'd need to declaratively state in `$NDK/meta`.

1. Clang location
2. AR location
3. Ranlib location
4. Set up yasm
5. Clang triple for the target
6. Additional cflags
7. Additional cxxflags
8. Additional ldflags
9. Additional exe linker flags
10. CMAKE_C_STANDARD_LIBRARIES_INIT
11. CMAKE_CXX_STANDARD_LIBRARIES_INIT
12. PIC/PIE options?
13. Sysroot location (for find_library and co)
14. Module location? (so we can expose gtest and co)

1-5 seem like exactly the intended use case of toolchain files (although there's the quirky piece about the clang target triple being derived from the ABI and the target OS version, and it sounds like what the CMake folks really don't want in toolchain files is any sort of logic?).

These aren't hard to communicate to CMake as tabular data as Brad requested, I'm just not sure why loading data, parsing it, and then setting the various compiler paths is preferable to just letting the NDK set those variables.

6-11 are to work around bugs where the Clang driver ought to use something by default but doesn't. They may be bugs, but bugs absolutely do happen and we will always need a way to fix them in the short term.

These are trickier (or at least bulkier) to specify because they vary by ABI, API level, and there's at least one case that needs to vary based on the presence of an ldflag (we can't currently handle this in the toolchain file because we don't know the ldflags used for each module). ABI/API variations are simple enough to handle by exposing plain old data, but again, an `if()` seems cleaner to me.

12 I'm not really sure about. This is referring to https://github.com/android-ndk/ndk/issues/575. This is specifically for ensuring that ndk-build and CMake both use the same PIC model.

13 and 14 I'm also not sure about. There's also the matter of disallowing the searching of the system's library directory because that's not something you want to do when cross compiling.

@Zingam @rcdailey @alexcohn: feel free to continue the discussion about the `cmake_minimum_version` stuff, but I'd recommend you do so over on https://issuetracker.google.com/110693527 where the right people are involved :) It seems like @rcdailey has already managed to convince Jomo though, so I think the only thing needed on that front is time.

**Comment by rcdailey** [comment]
@bradking 

> I'd like the same information to be available to all build systems without having to parse .cmake files.

What does this mean? `Android.cmake` has to be parsed whether it's external or not, right? I'm obviously severely lacking in my knowledge of how CMake handles platform-specific criteria, since almost your whole response confuses me :-(

**Comment by DanAlbert** [comment]
AIUI he's looking for the same information to be shared between CMake, ndk-build, bazel, etc.

> Still, it would be nice to re-use non-cmake-specific things where possible.

This isn't any different than what I'd planned, FWIW. Right now we generate things like `$NDK/build/cmake/abis.cmake` from `$NDK/meta/abis.json`. We can't read JSON from CMake files, so we preprocess it to handle that. We do the same thing for ndk-build. The same metadata files (the JSON ones) also feed gradle and it's what I expect people to use for other build systems.

But there are some things that are just not well expressed in a declarative way and it's much simpler for us to just implement it in each language. As an example, `Wl,--fix-cortex-a8`. We need that flag if we're linking ARM with gold or bfd. We need to avoid it if we're linking with LLD. We don't know if we should use that flag without scanning the flags that will be used to link the binary.

For a simpler example, we need to use `-mstackrealign` on x86 when the target API level is below 24. We could handle this by having the JSON files list cflags/cppflags/ldflags for the cross product of ABI/API, but it's really much simpler to just write the `if`. In a case like this where it's a bug that we need this flag at all, that seems preferable to me.

**Comment by DanAlbert** [comment]
My plan was to document it. There's a not-yet-landed doc I've written that is a guide for build system maintainers that will cover any of the issues that are difficult to express declaratively.

**Comment by DanAlbert** [comment]
Based on the discussion here I'm not sure who this work benefits. It makes our job a fair amount more difficult, reduces our ability to work around bugs, and doesn't seem to benefit users in any tangible way (aside from maybe clearing up the confusion about there being two options, but I suspect that by doing this I'd really just be adding a third).

There's very clearly some work to be done in terms of documentation here, so I'm going to leave this open to track that.

The part of this work that is factoring out pieces of the NDK metadata into the JSON files in `$NDK/meta` will continue as well, since that is still useful for sharing information between ndk-build, the CMake toolchain file, and the gradle plugin. Those are intended as stable API for the NDK (the information may be extended, but we will keep the information that exists), so other build systems are free to make use of that. If there's information we're missing that ought to be there, feel free to send a patch or open a bug. As long as it's something that's feasible for us to expose it's worth doing imo.

If I've misunderstood the opinions in this thread, please speak up.

**Comment by rcdailey** [comment]
@DanAlbert I think then it makes sense to remove the built in CMake support. Doing nothing means we have two distinctly different ways of cross compiling. One way that works and is officially supported (the toolchain file). Another done by Kitware that is demonstrably non-functional and not officially maintained.

**Comment by DanAlbert** [comment]
I agree but that's not in my control.

**Comment by bradking** [comment]
I've opened [CMake issue 18787](https://gitlab.kitware.com/cmake/cmake/issues/18787) to record the proposal to remove builtin support.

**Comment by bradking** [comment]
One way to at least partially reconcile things is to reduce the NDK toolchain file's dependence on undocumented non-public behavior of CMake.  In part that means creating `Platform/*.cmake` modules for Android over in the NDK and having the toolchain file point to them via `CMAKE_MODULE_PATH`.  An audit of the toolchain file's code would be needed to determine what is supposed to go where.  That approach would keep everything in the NDK and outside of CMake but still load the various settings at the right time.

**Comment by rcdailey** [comment]
To follow up to what @bradking suggested, which I agree with, I want to go back to an [older comment from @DanAlbert](https://github.com/android-ndk/ndk/issues/463#issuecomment-423631480):

> Yes, https://cmake.org/cmake/help/git-master/policy/CMP0017.html prevents this. When I asked Brad if he had any suggestions on how we could make an implementation like this work, he rejected the implementation idea I've proposed above. CMake won't approve any change that delegates work outside of CMake, so the only thing we're able to put in the NDK is plain old data. We also can't touch `CMAKE_MODULE_PATH`.

I think this still is a concern, but with the CMake issue brad filed above ([Issue 18787](https://gitlab.kitware.com/cmake/cmake/issues/18787)), this should no longer be a concern. Because right now CMake has their own android platform modules, any specified by the NDK would not be used. However if they get rid of it, my understanding of that policy is that the NDK's modules will instead be used.

**Comment by bradking** [comment]
Right, I forgot about CMP0017.  It exists for good reasons and will not be removed.  It's the presence of `Platform/Android*` modules in upstream CMake that break the possibility of the NDK using that namespace.  Even if not removed, CMake's copy of those modules could be taught to detect when they are being loaded by the NDK's toolchain file and directly `include()` copies found in the NDK.

**Comment by bradking** [comment]
If anyone wants to work on that approach, start by doing `rm -rf Modules/Platform/Android*` in your CMake installation and then see if you can get the NDK toolchain file to work as discussed above.  If such a solution can be demonstrated then we can consider upstream CMake changes in support of it.

**Comment by DanAlbert** [comment]
> Maybe, assuming CMake can keep from completely removing the functionality covered by CMP0017, until issue 18787 can be resolved, you can check for CMAKE_VERSION >= 2.8.4 and CMAKE_VERSION < (version that fixed 18787), and if true, disable CMP0017 and then modify CMAKE_MODULE_PATH? Something along those lines...

This is what I was going to suggest. I very much would prefer to implement this as if it were any other platform, so if there's a way for me to do this without disabling CMP0017 on new enough (as of yet unwritten) versions of CMake then I don't have a problem with forcibly disabling this behavior in the toolchain file for old versions (it's not ideal, but I don't know that we have another option).

If this form would be acceptable (this was one of the original attempts I'd tried and it wasn't then, but I don't think we were considering the upstream change to make disabling CMP0017 unnecessary at that point), then I'll put this bug back in its original form and we can take on the work (though I can't make any guesses as to when I'll be able to tackle it, so if someone else is looking to contribute don't let me stop you).

**Comment by rcdailey** [comment]
@bradking I made an edit to my last post, but would this work? It wouldn't require any further changes to CMake as long as you maintain deprecation status for CMP0017:

> Maybe, assuming CMake can keep from completely removing the functionality covered by CMP0017, until issue 18787 can be resolved, you can check for `CMAKE_VERSION >= 2.8.4` and `CMAKE_VERSION < (version that fixed 18787)`, and if `true`, disable CMP0017 and then modify `CMAKE_MODULE_PATH`? Something along those lines...

**Comment by DanAlbert** [comment]
> Technically you do, but it requires that you maintain both the toolchain and the platform scripts.

That's basically the same as status quo. I don't have to do anything if I want that :)

**Comment by rcdailey** [comment]
Well except that 1) you have to copy the platform scripts from CMake and place them in the NDK, and 2) you have to fix & maintain them to at least equivalent functionality provided by the toolchain.

I get that we're sort-of joking about it now, but the minute Brad says CMP0017 is or will be deleted, it won't be much of a joke anymore... :-(

**Comment by rcdailey** [comment]
well technically the escape hatch has always existed, e.g. the toolchain file. Because that existed long before CMP0017 was created, IIRC. But I'm not sure what criteria you guys are basing this on. Either way, doesn't matter, I think it should be contingent on the Module files moving to the NDK repository as well as ownership of maintenance of them before it's removed. But sounds like that's what you are saying should happen.

So in that case, I think my idea works for backward compatibility (Newer NDKs on older CMake versions that still have the Android platform files). Sounds like a plan to me.

**Comment by DanAlbert** [comment]
It sounds like we're all agreed on the form of this now (thanks everyone for the feedback; it's been very helpful). I've updated the title and added a summary to the OP to clarify.

I'm in the process of updating our roadmap right now. This currently sits at third on our priority list for major work. If I had to guess that lands it in r21 (Q3), but it's possible that the things ahead of it take more or less time than I expect. I'll slot this into a specific milestone when I'm more certain.

**Comment by Zingam** [comment]
@DanAlbert Is there a reason for AS to not update CMake regularly? Is their work unrelated to yours? What about removing CMake from AS and including it in the NDK as I cannot think about a scenario of using CMake without the NDK anyway?
I am glad that you have changed the title! 👍

**Comment by DanAlbert** [comment]
Yes, we're not the droids you're looking for. That's something the studio folks handle.

I worry that it would be harder to update if it we're in the NDK proper, actually. The IDE integration aspects seem to be quite fragile (see the bugs that we're opened when 3.10 was added to the SDK), so it's probably best to let the IDE engineers run that.

**Comment by DanAlbert** [comment]
(though if you haven't seen it, you don't need to wait for the SDK to have a new CMake now, you can set `cmake.dir` in your `local.properties` and use one that you've installed yourself)

**Comment by rcdailey** [comment]
There's a CMake version property you can add to your build.gradle to make
it use an external CMake (looks for it on PATH):

https://developer.android.com/studio/projects/add-native-code#use_a_custom_cmake_version

You can also use a '+' symbol to specify a minimum version instead of an
exact version. This is undocumented though. Please see the issue I filed
here:

https://issuetracker.google.com/issues/110693527

**Comment by jlk1980** [comment]
Droid is a nightmare to try and do this on!

**Comment by bradking** [comment]
> We won't remove support for CMP0017 OLD in a version of CMake that doesn't have the escape hatch in place.

[CMake MR 4372](https://gitlab.kitware.com/cmake/cmake/merge_requests/4372) demonstrates such an escape hatch.  In that proposal, the Android NDK toolchain file could set `CMAKE_ANDROID_NDK_HOOK_DIR` to point at a location in the NDK that contains `Platform/Android*.cmake` modules.  Doing so tells CMake's builtin modules of the same name to simply include the NDK copies and exit.

**Comment by bradking** [comment]
OTOH, as of CMake 3.16.4's support for the NDK r19+ unified toolchain, much of CMake's builtin knowledge of Android flags and such is no longer needed.  I'd like to again propose that CMake's upstream solution become the official way to build for Android.  It can even be used by the NDK's toolchain file.  We just need to add a few hooks to CMake's modules to load information from the NDK so that the NDK remains in control of the flags.

**Comment by DanAlbert** [comment]
@bradking Thanks for the update and sorry for not having had the time to resolve this yet.

The proposed change is what I had in mind, and I'd played a bit with something similar back when I was first looking at this with good results.

Agreed that simpler is better. We'll have a look at exactly what's left for the r19+ toolchains. We might be able to do something simpler there, but won't know until we look. Not sure exactly when that will be. Might be something @hhb can look at soon.

**Comment by enh-google** [comment]
...and in the meantime, if you have specific examples of things we could do that would help make even more of "CMake's builtin knowledge of Android flags and such ... no longer needed", let us know.

**Comment by bradking** [comment]
> if you have specific examples of things we could do that would help make even more of "CMake's builtin knowledge of Android flags and such ... no longer needed",

@enh-google off the top of my head the main remaining problem is all the abi-specific flags like `-march=armv7-a` and `-mstackrealign`.  The Android-specific toolchains should be able to know all the Android ABI conventions based only on the target triple.  Build systems should not have to all memorize that information.  If the toolchain were to apply such settings automatically then they would be maintained inside the NDK regardless of what build system is used.

**Comment by DanAlbert** [comment]
`-mstackrealign` already is managed by clang, I thought. Maybe I'm misremembering.

We do need to find some time to finish pushing things into the driver though.

**Comment by DanAlbert** [comment]
(I'm someone can start working on this for r23, but keeping in unplanned until we actually have someone free for it)

**Comment by ericoporto** [comment]
Any chance this is picked up?

**Comment by DanAlbert** [comment]
Not yet. I think it's what @hhb is looking at after LLDB.

**Comment by hhb** [comment]
To make sure everyone is on the same page: currently there are two ways to build for Android using cmake.
1) Using toolchain file provided by NDK: `-DCMAKE_TOOLCHAIN=/path/to/android.toolchain.cmake`
2) Using cmake upstream support: `-DCMAKE_SYSTEM_NAME=Android`

They have totally different implementation and depend on each other's internal. The goal of this work is to merge the implementation:
1) android.toolchain.cmake will become a layer on top of cmake upstream support. It will provide the ANDROID_* interface.
2) some mechanism is needed for NDK to hook into cmake, to provide additional information.

Since android.toolchain.cmake will be built on top of cmake Android support. As long as cmake interface is stable (likely), thing should work. On the other hand, if a NDK bug requires a change on cmake, people will be forced to upgrade their cmake. That's what we should try to avoid.

I plan to add hooks similar to [cmake1112](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/1112). For each related cmake module, we add both pre and post hook like:

```cmake
if(CMAKE_ANDROID_NDK)
  include(${CMAKE_ANDROID_NDK}/build/cmake/hooks/{pre,post}/Android-Determine.cmake OPTIONAL)
endif()
```

(See also [cmake4372](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4372), [cmake1092](https://gitlab.kitware.com/cmake/cmake/-/merge_requests/1092), [ndk395513](https://android-review.googlesource.com/c/platform/ndk/+/395513))

In most cases, the hooks will only add cxx / linker flags. Because flags are more frequently changed on NDK side. It is easier to manage them in NDK. But we can also use the hooks to change cmake behaviors. Some examples:

[e043c26](https://android.googlesource.com/platform/ndk/+/e043c269d3c0090954f2f405bfe9f6bffcdf6b9a) Switch to llvm-strip, advocate its use.
[48b8e52](https://android.googlesource.com/platform/ndk/+/48b8e52eebbcef8f727dd081d13c88e9e7c7d35a) Switch to llvm-ar
cmake sets _CMAKE_TOOLCHAIN_PREFIX and prefers ${triple}-ar. With hooks, NDK can set CMAKE_AR / CMAKE_STRIP, until triple prefixed binutils are removed, or some fix is applied upstream in cmake.

[f563d92](https://android.googlesource.com/platform/ndk/+/f563d927f570b8d025c2f4a837533eba2363ca74) Fix find_library search order.
This will have to be fixed in cmake. But we may hack CMAKE_SYSTEM_*_PATH in a hook for older cmake.

**Comment by DanAlbert** [comment]
And a further note that wherever possible we'll fix those extra flags directly in Clang rather than needing to patch them into CMake like this. Ideally those files will be a no-op when using an up to date CMake. Toolchain updates take a long time though so that's not always practical.

**Comment by DanAlbert** [comment]
The NDK side of this merged (for r23) yesterday. It requires CMake 3.20.
